### PR TITLE
sink(ticdc): actively checks the size of min.insync.replicas and replication-factor

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflow/issues/3994

### What is changed and how it works?

actively checks the size of min.insync.replicas and replication-factor.

Error reported when min.insync.replicas is less than replication-factor.

### Check List 

Tests 

 - [x] Unit test
 - [x] Manual test See: https://github.com/pingcap/tiflow/pull/4263#issuecomment-1009638808

Code changes

 - Has exported function/method change

Side effects

None

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Fix the problem that TiCDC cannot send messages when `min.insync.replicas` is less than `replication-factor`
```

### CODE REVIEW

**Comment by hi-rustin** [comment]
/CC @overvenus  @amyangfei 
I'm doing local manual testing, but the PR is ready for review.
Could you please take a look? Thanks! 💚 💙 💜 💛 ❤️

File: cdc/sink/producer/kafka/kafka.go [hunk 1]
```diff
@@ -428,43 +445,81 @@ func validateMaxMessageBytesAndCreateTopic(admin kafka.ClusterAdminClient, topic
...
+		}
+
+		if replicationFactor < minInsyncReplicas {
+			msg := fmt.Sprintf("`replication-factor` cannot be smaller than the `%s` of topic",
+				kafka.MinInsyncReplicasConfigName)
+			log.Error(msg, zap.Int("replicationFactor", replicationFactor),
+				zap.Int("minInsyncReplicas", minInsyncReplicas))
+			return errors.New(msg)
+		}
+	}
```
Comment (amyangfei) [inline cdc/sink/producer/kafka/kafka.go:470]: What about extracting a function, and reuse it in L472-L487
```golang
minInsyncReplicasStr, err := minInsyncReplicasConfigGetter(...)
minInsyncReplicas, err := strconv.Atoi(minInsyncReplicasStr)
if replicationFactor < minInsyncReplicas {
    //...
}
```

File: cdc/sink/producer/kafka/kafka.go [hunk 2]
```diff
@@ -428,43 +445,81 @@ func validateMaxMessageBytesAndCreateTopic(admin kafka.ClusterAdminClient, topic
...
+	if exists {
+		minInsyncReplicasStr, err := getTopicConfig(admin, info,
+			kafka.MinInsyncReplicasConfigName,
+			kafka.MinInsyncReplicasConfigName)
+		if err != nil {
+			return err
+		}
+		minInsyncReplicas, err := strconv.Atoi(minInsyncReplicasStr)
+		if err != nil {
+			return err
```
Comment (amyangfei) [inline cdc/sink/producer/kafka/kafka.go:460]: Better to use an rfc error here.

File: cdc/sink/producer/kafka/kafka.go [hunk 3]
```diff
@@ -428,43 +445,81 @@ func validateMaxMessageBytesAndCreateTopic(admin kafka.ClusterAdminClient, topic
...
+	if err != nil {
+		return err
+	}
+
+	if replicationFactor < minInsyncReplicas {
+		msg := fmt.Sprintf("`replication-factor` cannot be smaller than the `%s` of broker",
+			kafka.MinInsyncReplicasConfigName)
+		log.Error(msg, zap.Int("replicationFactor", replicationFactor),
+			zap.Int("minInsyncReplicas", minInsyncReplicas))
+		return errors.New(msg)
```
Comment (amyangfei) [inline cdc/sink/producer/kafka/kafka.go:486]: ditto

File: cdc/sink/producer/kafka/kafka.go (hunk 2 above)
Comment (hi-rustin) [reply cdc/sink/producer/kafka/kafka.go:460]: Warped in https://github.com/pingcap/tiflow/blob/4b4f59676601b25d8c17fbb8564c7d781b2d779f/cdc/sink/producer/kafka/kafka.go#L357

File: cdc/sink/producer/kafka/kafka.go (hunk 1 above)
Comment (hi-rustin) [reply cdc/sink/producer/kafka/kafka.go:470]: Good idea!

Comment (hi-rustin) [reply cdc/sink/producer/kafka/kafka.go]: Good idea! Addressed!

**Review by 3AceShowHand — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

**Comment by hi-rustin** [comment]
/hold

Waiting for testing.

**Comment by hi-rustin** [comment]
Scenario 1: Both values are 1 
1. Start the kafka cluster using the following configuration and commands.
```yaml
version: '3'

networks:
  ticdc-test:
    driver: bridge

services:
  zookeeper:
    image: wurstmeister/zookeeper
    networks:
      - ticdc-test
    ports:
     - "2181:2181"

  kafka:
    image: wurstmeister/kafka
    networks:
      - ticdc-test
    ports:
     - "9092"
    environment:
      DOCKER_API_VERSION: 1.22
      KAFKA_ADVERTISED_HOST_NAME: 172.23.29.101
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_MIN_INSYNC_REPLICAS: 1
      KAFKA_DEFAULT_REPLICATION_FACTOR: 3
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 3
      KAFKA_CREATE_TOPICS: "ticdc-test:1:1"
      KAFKA_MESSAGE_MAX_BYTES: 3145728
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    depends_on:
      - zookeeper
```
```console
docker-compose -f ./docker-compose-kafka.yaml up --scale kafka=3 -d
```
2. Start the tidb and ticdc clusters.
3. Create a changefeed using an existing topic.
```console
./cdc cli changefeed create   --sink-uri="kafka://127.0.0.1:64324/ticdc-test?kafka-version=2.6.0&protocol=canal-json"
```
4. Check the normal output of the message.

Scenario 2: When the topic exists and the replication factor is less than the minimum synchronization requirement.
1. Start the kafka cluster using the following configuration and commands.
```yaml
version: '3'

networks:
  ticdc-test:
    driver: bridge

services:
  zookeeper:
    image: wurstmeister/zookeeper
    networks:
      - ticdc-test
    ports:
     - "2181:2181"

  kafka:
    image: wurstmeister/kafka
    networks:
      - ticdc-test
    ports:
     - "9092"
    environment:
      DOCKER_API_VERSION: 1.22
      KAFKA_ADVERTISED_HOST_NAME: 172.23.29.101
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_MIN_INSYNC_REPLICAS: 2
      KAFKA_DEFAULT_REPLICATION_FACTOR: 3
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 3
      KAFKA_CREATE_TOPICS: "ticdc-test:1:1"
      KAFKA_MESSAGE_MAX_BYTES: 3145728
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    depends_on:
      - zookeeper

```
```console
docker-compose -f ./docker-compose-kafka.yaml up --scale kafka=3 -d
```
2. Start the tidb and ticdc clusters.
3. Create a changefeed using an existing topic.
```console
./cdc cli changefeed create   --sink-uri="kafka://127.0.0.1:41396/ticdc-test?kafka-version=2.6.0&protocol=canal-json"
```
4. check the cli output 
```log
      --pd string          PD address, use ',' to separate multiple PDs (default "http://127.0.0.1:2379")

[CDC:ErrKafkaNewSaramaProducer]new sarama producer: [CDC:ErrKafkaInvalidConfig]because TiCDC Kafka producer's `request.required.acks` defaults to -1, TiCDC cannot deliver messages when the `replication-factor` is less than `min.insync.replicas`: `replication-factor` cannot be smaller than the `min.insync.replicas` of topic
```

Scenario 2: When the topic does not exist and the replication factor is less than the minimum synchronization requirement.
1. Start the kafka cluster using the following configuration and commands.
```yaml
version: '3'

networks:
  ticdc-test:
    driver: bridge

services:
  zookeeper:
    image: wurstmeister/zookeeper
    networks:
      - ticdc-test
    ports:
     - "2181:2181"

  kafka:
    image: wurstmeister/kafka
    networks:
      - ticdc-test
    ports:
     - "9092"
    environment:
      DOCKER_API_VERSION: 1.22
      KAFKA_ADVERTISED_HOST_NAME: 172.23.29.101
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_MIN_INSYNC_REPLICAS: 2
      KAFKA_DEFAULT_REPLICATION_FACTOR: 3
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 3
      KAFKA_CREATE_TOPICS: "ticdc-test:1:1"
      KAFKA_MESSAGE_MAX_BYTES: 3145728
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    depends_on:
      - zookeeper

```
```console
docker-compose -f ./docker-compose-kafka.yaml up --scale kafka=3 -d
```
2. Start the tidb and ticdc clusters.
3. Create a changefeed using an existing topic.
```console
./cdc cli changefeed create   --sink-uri="kafka://127.0.0.1:41396/ticdc-test2?kafka-version=2.6.0&protocol=canal-json"
```
4. check the cli output 
```log
      --pd string          PD address, use ',' to separate multiple PDs (default "http://127.0.0.1:2379")

[CDC:ErrKafkaNewSaramaProducer]new sarama producer: [CDC:ErrKafkaInvalidConfig]because TiCDC Kafka producer's `request.required.acks` defaults to -1, TiCDC cannot deliver messages when the `replication-factor` is less than `min.insync.replicas`: `replication-factor` cannot be smaller than the `min.insync.replicas` of broker

```

**Pull request merged**